### PR TITLE
Integrate Google Cloud MemoryStore for Redis into gcp-service-broker

### PR DESCRIPTION
**Pull request by sophiawho**
Add Redis as a built-in service with two service plans corresponding to the two service tiers:
- basic
- standard high availability

Users can also specify the Redis instance's
- authorized network
- capacity tier
- display name
- instance id
- location id

Service Broker outputs instance variables to client can connect to the Redis instance: 
- name
- host
- port
- redis version
- memorysize in GiB

Console output after creating an instance example with "Basic Tier" examples:
```
2019/06/21 12:08:18 Running Example: google-memorystore-redis/Basic Redis Configuration
2019/06/21 12:08:18 gcp-service-broker client provision --instanceid "ex1103823129" --planid "dd1923b6-ac26-4697-83d6-b3a0c05c2c94" --serviceid "3ea92b54-838c-4fe1-b75d-9bda513380aa" --params "{}"
2019/06/21 12:08:18 gcp-service-broker client bind --instanceid "ex1103823129" --planid "dd1923b6-ac26-4697-83d6-b3a0c05c2c94" --serviceid "3ea92b54-838c-4fe1-b75d-9bda513380aa" --bindingid "ex1103823129" --params "{\"role\":\"redis.viewer\"}"
2019/06/21 12:08:18 gcp-service-broker client unbind --instanceid "ex1103823129" --planid "dd1923b6-ac26-4697-83d6-b3a0c05c2c94" --serviceid "3ea92b54-838c-4fe1-b75d-9bda513380aa" --bindingid "ex1103823129"
2019/06/21 12:08:18 gcp-service-broker client deprovision --instanceid "ex1103823129" --planid "dd1923b6-ac26-4697-83d6-b3a0c05c2c94" --serviceid "3ea92b54-838c-4fe1-b75d-9bda513380aa"
2019/06/21 12:08:18 Provisioning google-memorystore-redis/Basic Redis Configuration
2019/06/21 12:12:30 PUT http://user:pass@localhost:8000/v2/service_instances/ex1103823129?accepts_incomplete=true -> 201, "{}\n"
2019/06/21 12:12:30 Binding google-memorystore-redis/Basic Redis Configuration
2019/06/21 12:12:33 PUT http://user:pass@localhost:8000/v2/service_instances/ex1103823129/service_bindings/ex1103823129 -> 201, "{\"credentials\":{\"Email\":\"gsb-binding-ex110382@jlewisiii-gsb.iam.gserviceaccount.com\",\"Name\":\"gsb-binding-ex110382\",\"PrivateKeyData\":\"ewogICJ0eXBlIjogInNlcnZpY2VfYWNjb3VudCIsCiAgInByb2plY3RfaWQiOiAiamxld2lzaWlpLWdzYiIsCiAgInByaXZhdGVfa2V5X2lkIjogIjFmNmZmYzZlODczMmYyMWE0N2UzN2EyMzNjMGEyMWNmNGEyYjdhNTMiLAogICJwcml2YXRlX2tleSI6ICItLS0tLUJFR0lOIFBSSVZBVEUgS0VZLS0tLS1cbk1JSUV2QUlCQURBTkJna3Foa2lHOXcwQkFRRUZBQVNDQktZd2dnU2lBZ0VBQW9JQkFRRExPUkE3Y0NWVHFjT1FcbjkwV2pzSUZka1YveTFzK3N2Z3B1RWRSYThMbC9JYkdrUmJnNjB6eWpNZkM0TlBOOWRmVFRhcllSTkxxSElJMENcbk4xekJFemVMY3gyZGd1ZHlpcjJRa3Zlb1BpRWFCeExnQVFIdzArNkF5MEgveEpEWnFJODBMZ3JHZXlKdEdkeXlcbnNWbjZ2dE00VUJDdkhSVm1VZmx1ZmZVaDJTWEZTTThEMUdkcTY2bnR6SlU5U3Z5NTNBZ2F1emZBS1ZuZXlvR3hcbmhLdndxUU9HbUR5cVVIRWl4YzlKc0RTcFIraU1KSFd0Z2cyRkZSZ1l2MzV4cnh2cVlObTg1aXM3ajZIbTNoc3pcblVmTFBIVFgwUWNKZkRGREhuUlh6SkxkUno1ei9PVlBVSVRXaU9qalloY1RrSnVzVlBpQVdPRGJwSjR0S0cwNFJcbkRIQ2QycGg1QWdNQkFBRUNnZ0VBQUppKzRJbk1Wa0gybDNKSkU2UUtSb2NpMHJ0STVLU0hNbnBlRnY2WXhpa3lcbnYyZk9xaFRpTUp3amdvcng3R2IxckhtUSszSnpSTC9MOTRYSVFZU0lBbmU0bzhuVldrdW9KdjN1STdjN3BpV3FcbkFNWGRVVmdjbmJ5d1h0V2ZLT3NhMjkxSEUxSkdaWVJ3cTl2NDhXaWJoVkZqaHliQ1NyNElDZDF0VHVlTXVleE5cbmlpMiswMTZtdEFIdy9uTVVWWWtmUERSY2xROFVCOU43NHpyT3J1dUJDSHJoQ0dlSUZBRi9YUHY5MnhDZ0R3SG5cbmsweUFSZU9VeDNrbGZ4NEo0dmR4NEkzSEprbkZZMU9UYzkrQzV3V0l0L3hyR0VjYTlXQlBqTjlMdUhsdzk4MmxcbnE3T080YUppK3JxVDBMU2Nkb2pwcHNLK2tnblBPTC9vNGl1M0JGMVVQUUtCZ1FEN0FVUlM0SFoyRU9qR1RadkFcbmJrRGlQMlVsclBTQjRsYlpSb1A3ck1EZlVveHFXWVk2b1ZTNFhiV3ZpQkV4L1kyTmZzNFRJYkdRUkdJM1BKZ1NcblMxMWp1OVVxbU1KTlo1TzF2TUh4VFZjUHB3bUpzTUtrcFNWdm5FZjYxT0lVbVJndzRLc1RWT0tDaklFTnhlMDRcbnRXNHIzVWNvZzFZRXRDOU9iOWdKNEM0MzlRS0JnUURQUkYrRFhua1kwaW8xaHBldnExaDhyV3I1NW14RlVpS29cbjZPbjNCK1g4aUd5MGpzcXQrV255UDI5NnZsZXk4WWp3SGIrKzR0eTF4TGc3YjgzMy9GNTR4eGkyN2hQeEkyMUlcbkpuTjZiRDBnWmhnU2dKQ0pkTzN2ZmZ1THJaczdrcWcySElMKzVtNUtDMnFMOGhYK2RYeU5xSUZOTjc3R2VySFdcbkRwMGxWUnIvOVFLQmdINUpJK1FDdFBjd0NMMkUzSUVUdDZkZE4vbWRyQ01SQ3Bza3BGRGltT09ETjh4bUlHZXRcbk5kL1FuaEdqaHVzODQ1UTJJVjUwekU2R2FWZmJMN084b1U4RmVQbUdnL1BIMGdSVTVNc1FHMlp2cHhmVWpBbWtcbllWSm5NYnZja2FiRlkrdHJqZ2NhNVIvRHEwYVB0NHVabE1XUy9qTSs2TzlTN0o0dEhXZlNsODVsQW9HQVQ2VEhcblN2U2ZkZEUwVlBNT0VIdTMrZGl0ZGE5OFJUcmxoS0JUWHhCUzZMSUhpMjRmSmJEa2JDakNEcGxibTdCQ1hYb2dcbjl4bTNrOFV1d2ZBWlprUThqUTU4U2JhOFJQbkhBV1Y5RkRySlZrekV6VDlIeXZuNmYvK2FERGMrb2lxZHBvVllcbk5OaVk0cXoyV1NaQ0cxQkw5eWNiNEc4ZVNLaFc4NWtUKzRkelk0MENnWUFIcDFXU0hWcTUyYzVhK0kvMVJ0ay9cbnpsRG91Z3BHbVh5QmhVS0ZVTVNMWEkwNkhBZllHVEdQZzVzTktDN0pINTJxbXdEMFpMa2x3dFdyWkJGOGRlbkRcbjY5bkgxNzBiZ1R0V3pHejcrcDFVMGhXY1BhVUZBQURoM01KdU4yTUVNcjdwMUlPOXllQjRrUDdYM1BsMXU1RzVcbjZjS1U5UFhTK01vdEFDa1A4a25kdUE9PVxuLS0tLS1FTkQgUFJJVkFURSBLRVktLS0tLVxuIiwKICAiY2xpZW50X2VtYWlsIjogImdzYi1iaW5kaW5nLWV4MTEwMzgyQGpsZXdpc2lpaS1nc2IuaWFtLmdzZXJ2aWNlYWNjb3VudC5jb20iLAogICJjbGllbnRfaWQiOiAiMTA1MTQwOTYyMjgzMzM4MTAwMDQ0IiwKICAiYXV0aF91cmkiOiAiaHR0cHM6Ly9hY2NvdW50cy5nb29nbGUuY29tL28vb2F1dGgyL2F1dGgiLAogICJ0b2tlbl91cmkiOiAiaHR0cHM6Ly9vYXV0aDIuZ29vZ2xlYXBpcy5jb20vdG9rZW4iLAogICJhdXRoX3Byb3ZpZGVyX3g1MDlfY2VydF91cmwiOiAiaHR0cHM6Ly93d3cuZ29vZ2xlYXBpcy5jb20vb2F1dGgyL3YxL2NlcnRzIiwKICAiY2xpZW50X3g1MDlfY2VydF91cmwiOiAiaHR0cHM6Ly93d3cuZ29vZ2xlYXBpcy5jb20vcm9ib3QvdjEvbWV0YWRhdGEveDUwOS9nc2ItYmluZGluZy1leDExMDM4MiU0MGpsZXdpc2lpaS1nc2IuaWFtLmdzZXJ2aWNlYWNjb3VudC5jb20iCn0K\",\"ProjectId\":\"jlewisiii-gsb\",\"UniqueId\":\"105140962283338100044\",\"host\":\"10.0.0.3\",\"memory_size_gb\":4,\"port\":\"6379\",\"redis_version\":\"REDIS_4_0\"}}\n"
2019/06/21 12:12:33 Unbinding google-memorystore-redis/Basic Redis Configuration
2019/06/21 12:12:34 DELETE http://user:pass@localhost:8000/v2/service_instances/ex1103823129/service_bindings/ex1103823129?service_id=3ea92b54-838c-4fe1-b75d-9bda513380aa&plan_id=dd1923b6-ac26-4697-83d6-b3a0c05c2c94 -> 200, "{}\n"
2019/06/21 12:12:34 Deprovisioning google-memorystore-redis/Basic Redis Configuration
2019/06/21 12:14:54 DELETE http://user:pass@localhost:8000/v2/service_instances/ex1103823129?accepts_incomplete=true&service_id=3ea92b54-838c-4fe1-b75d-9bda513380aa&plan_id=dd1923b6-ac26-4697-83d6-b3a0c05c2c94 -> 200, "{}\n"
2019/06/21 12:14:54 Cleaning up the environment
2019/06/21 12:14:54 Unbinding google-memorystore-redis/Basic Redis Configuration
2019/06/21 12:14:54 DELETE http://user:pass@localhost:8000/v2/service_instances/ex1103823129/service_bindings/ex1103823129?service_id=3ea92b54-838c-4fe1-b75d-9bda513380aa&plan_id=dd1923b6-ac26-4697-83d6-b3a0c05c2c94 -> 410, "{}\n"
2019/06/21 12:14:54 Deprovisioning google-memorystore-redis/Basic Redis Configuration
2019/06/21 12:14:54 DELETE http://user:pass@localhost:8000/v2/service_instances/ex1103823129?accepts_incomplete=true&service_id=3ea92b54-838c-4fe1-b75d-9bda513380aa&plan_id=dd1923b6-ac26-4697-83d6-b3a0c05c2c94 -> 410, "{}\n"
2019/06/21 12:14:54 Success

```

### CODE REVIEW

File: pkg/providers/builtin/redis/broker.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
+// Copyright 2018 the Service Broker Project Authors.
```
Comment (josephlewis42) [inline pkg/providers/builtin/redis/broker.go]:
```suggestion
// Copyright 2019 the Service Broker Project Authors.
```

**Comment by sophiawho** [comment]
All authors have signed off and are ok with their commits being contributed this project.

File: pkg/providers/builtin/redis/broker.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+		return models.ServiceInstanceDetails{}, fmt.Errorf("Error creating new Redis instance: %s", err)
+	}
+
+	resp, err := op.Wait(ctx)
+	if err != nil {
+		return models.ServiceInstanceDetails{}, err
+	}
+
+	id := models.ServiceInstanceDetails{
+		Name: resp.Name,
```
Comment (josephlewis42) [inline pkg/providers/builtin/redis/broker.go]: I think we want a few more fields here, probably the IP address/host name, version of Redis, the port number, and the size of the instance (copy of memory_size_gb) so the application that this is bound to can connect to it.

Comment (josephlewis42) [reply pkg/providers/builtin/redis/broker.go]: You can check out how cloud-storage does this with the InstanceInformation struct to store additional fields beyond what's available in ServiceInstanceDetails

File: pkg/providers/builtin/redis/broker.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+var serviceTiers = map[string]redis.Instance_Tier{
+	"basic": redis.Instance_BASIC,
+	"standard_ha": redis.Instance_STANDARD_HA,
+}
+
+// Provision creates a new Redis instance from the settings in the user-provided details and service plan.
+func (b *RedisBroker) Provision(ctx context.Context, provisionContext *varcontext.VarContext) (models.ServiceInstanceDetails, error) {
+
+	authorizedNetwork := provisionContext.GetString("authorized_network")
+	capacityTier := int32(provisionContext.GetInt("capacity_tier"))
```
[the remark this answers is not in the document]
Comment (sophiawho) [reply pkg/providers/builtin/redis/broker.go]: Done

File: pkg/providers/builtin/redis/broker.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	displayName := provisionContext.GetString("display_name")
+	instanceId := provisionContext.GetString("instance_id")
+	locationId := provisionContext.GetString("location_id")
```
[the remark this answers is not in the document]
Comment (sophiawho) [reply pkg/providers/builtin/redis/broker.go]: Done

File: pkg/providers/builtin/redis/broker.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+	serviceTier := serviceTiers[provisionContext.GetString("service_tier")]
```
[the remark this answers is not in the document]
Comment (sophiawho) [reply pkg/providers/builtin/redis/broker.go]: good idea, done

**Comment by josephlewis42** [comment]
The CLA has been signed and is consent is given, GitHub suggested comments is using the wrong account for suggested code.

**Pull request closed without merging**